big_tests/tests/mongoose_helper.erl

```diff
@@ -266,22 +266,58 @@ logout_user(Config, User) ->
             end
     end.
 
+% @doc Waits for `Fun` to return `ExpectedValue`.
+% Each time Different value is returned or
+% functions throws an error, we sleep.
+% Sleep time is 2x longer each try
+% but not longer than `max_time`
+factor_backoff(Fun, ExpectedValue, #{attempts := Attempts,
+                                     min_time := Mintime,
+                                     max_time := Maxtime}) ->
+    do_wait_until(Fun, ExpectedValue, #{attempts => Attempts,
+                                        sleep_time => Mintime,
+                                        history => [],
+                                        next_sleep_time_fun => fun(E) ->
+                                                                  min(E * 2, Maxtime)
+                                                          end}).
+
 wait_until(Predicate, Attempts, Sleeptime) ->
-     wait_until(Predicate, true, Attempts, Sleeptime).
-
- wait_until(Fun, ExpectedValue, Attempts, SleepTime) ->
-     wait_until(Fun, ExpectedValue, Attempts, SleepTime, []).
-
- wait_until(_Fun, _ExpectedValue, 0, _SleepTime, History) ->
-     error({badmatch, History});
- wait_until(Fun, ExpectedValue, AttemptsLeft, SleepTime, History) ->
-     case Fun() of
-         ExpectedValue ->
-             ok;
-         OtherValue ->
-             timer:sleep(SleepTime),
-             wait_until(Fun, ExpectedValue, dec_attempts(AttemptsLeft), SleepTime, [OtherValue | History])
-     end.
+    wait_until(Predicate, true, Attempts, Sleeptime).
+
+wait_until(Fun, ExpectedValue, Attempts, SleepTime) ->
+    do_wait_until(Fun, ExpectedValue, #{attempts => Attempts,
+                                        sleep_time => SleepTime,
+                                        history => [],
+                                        next_sleep_time_fun => fun(E) -> E end}).
+
+
+do_wait_until(_Fun, _ExpectedValue, #{attempts := 0, history := History}) ->
+    error({badmatch, History});
+do_wait_until(Fun, ExpectedValue, #{attempts := AttemptsLeft,
+                                    sleep_time := SleepTime,
+                                    history := History,
+                                    next_sleep_time_fun := NextSleepTimeFun} = State) ->
+    try Fun() of
+        ExpectedValue ->
+            ok;
+        OtherValue ->
+            wait_and_continue(Fun, ExpectedValue, OtherValue, State)
+    catch Error:Reason ->
+              wait_and_continue(Fun, ExpectedValue, {Error, Reason}, State)
```

Comment:
@arcusfelis, @michaltomczuk, I don't think so, but - do we need a tail call here? In theory `wait_until` can take as `infinity` as an `attempts` value.  Maybe we should resign from `infinity` thing? Especially that it is not used anywhere?

Reply:
@michalwski instead of @michaltomczuk, sorry :p

Reply:
@kanes115 I think there is no point in allowing `infinity` here. Tail recursion is not very needed here as this unlikely to have many recursive calls. On the other hand tail recursion in the `try ... of ... catch` is rather easy to achieve. Take a look at https://stackoverflow.com/questions/17559946/no-tail-recursive-code-in-a-try-catch-block

Reply:
`infinity` can not work well with heat death of the universe.
Also, debugging tests stuck in the retry loop is _not cool_.
Max time of execution for the whole call should be not more than 30 seconds.

Reply:
btw, this function is tail recursive.

```erlang
my_fun() ->
try ... of
x -> my_fun()
catch _:_ ->
error
end.
```

Because try does not catch errors in `of` block.

Reply:
Thanks!

Reply:
Right, looks like it's ready to be merged. @arcusfelis do you have any other comments?